fuel-types/src/canonical.rs

```diff
@@ -0,0 +1,619 @@
+                // The resulting value is always well-aligned.
+                let bytes = <$t>::to_be_bytes(*self);
+                for _ in 0..alignment_bytes(bytes.len()) {
+                    // Zero-pad
+                    buffer.push_byte(0)?;
+                }
+                buffer.write(bytes.as_ref())?;
+                Ok(())
+            }
+        }
+
+        impl Deserialize for $t {
+            const UNALIGNED_BYTES: bool = $unpadded;
+
+            fn decode_static<I: Input + ?Sized>(buffer: &mut I) -> Result<Self, Error> {
+                let mut asset = [0u8; ::core::mem::size_of::<$t>()];
+                buffer.skip(alignment_bytes(asset.len()))?; // Skip zero-padding
+                buffer.read(asset.as_mut())?;
+                Ok(<$t>::from_be_bytes(asset))
+            }
+        }
+    };
+}
+
+impl_for_primitives!(u8, true);
+impl_for_primitives!(u16, false);
+impl_for_primitives!(u32, false);
+impl_for_primitives!(u64, false);
+impl_for_primitives!(u128, false);
+
+impl SerializedSize for () {
+    const SIZE_STATIC: usize = 0;
+}
+
+// Empty tuple `()`, i.e. the unit type takes up no space.
+impl Serialize for () {
+    const SIZE_NO_DYNAMIC: bool = true;
+
+    #[inline(always)]
+    fn size_static(&self) -> usize {
+        0
+    }
+
+    #[inline(always)]
+    fn size_dynamic(&self) -> usize {
+        0
+    }
+
+    #[inline(always)]
+    fn size(&self) -> usize {
+        0
+    }
+
+    #[inline(always)]
+    fn encode_static<O: Output + ?Sized>(&self, _buffer: &mut O) -> Result<(), Error> {
+        Ok(())
+    }
+}
+
+impl Deserialize for () {
+    fn decode_static<I: Input + ?Sized>(_buffer: &mut I) -> Result<Self, Error> {
+        Ok(())
+    }
+}
+
+/// To protect against malicious large inputs, vectors size is limited on decoding.
+pub const VEC_DECODE_LIMIT: usize = 100 * (1 << 20); // 100 MiB
+
+impl<T: Serialize> SerializedSize for Vec<T> {
+    const SIZE_STATIC: usize = 8;
+}
+
+impl<T: Serialize> Serialize for Vec<T> {
+    const SIZE_NO_DYNAMIC: bool = false;
+
+    #[inline(always)]
+    // Encode only the size of the vector. Elements will be encoded in the
+    // `encode_dynamic` method.
+    fn encode_static<O: Output + ?Sized>(&self, buffer: &mut O) -> Result<(), Error> {
+        assert!(
+            self.len() < VEC_DECODE_LIMIT,
+            "Refusing to encode vector too large to be decoded"
+        );
+        let len: u64 = self.len().try_into().expect("msg.len() > u64::MAX");
+        len.encode(buffer)
+    }
+
+    fn encode_dynamic<O: Output + ?Sized>(&self, buffer: &mut O) -> Result<(), Error> {
+        // Bytes - Vec<u8> it a separate case without padding for each element.
+        // It should padded at the end if is not % ALIGN
+        if T::UNALIGNED_BYTES {
+            // SAFETY: `UNALIGNED_BYTES` only set for `u8`.
+            let bytes = unsafe { ::core::mem::transmute::<&Vec<T>, &Vec<u8>>(self) };
```

Comment:
Do we have any safety checks to ensure we aren't transmuting heap pointers here like `Box` etc?

Reply:
The `Box` works with raw pointers, so the memory layout for any `T` is the same. But maybe it is better to work directly with `&[u8]` to avoid any question in the future=)

Reply:
`T::UNALIGNED_BYTES` is only set for `u8` as the comment says. Types are also serialized bottom-up so this shouldn't be too much of an issue anyway. I'm still considering removing the `unsafe` usage, since I'm not sure that the performance impact will be meaningful.